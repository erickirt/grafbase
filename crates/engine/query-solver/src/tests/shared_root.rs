use crate::assert_solving_snapshots;

const SCHEMA: &str = r###"
enum join__Graph {
  CATEGORY @join__graph(name: "category", url: "http://localhost:4200/shared-root/category")
  NAME @join__graph(name: "name", url: "http://localhost:4200/shared-root/name")
  PRICE @join__graph(name: "price", url: "http://localhost:4200/shared-root/price")
}

type Product
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
  nested: Nested!
  category: String @join__field(graph: CATEGORY)
  name: String @join__field(graph: NAME)
  price: Float @join__field(graph: PRICE)
}

type Nested
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  id: ID!
}

type Query
  @join__type(graph: CATEGORY)
  @join__type(graph: NAME)
  @join__type(graph: PRICE)
{
  product: Product
  products: [Product]
}
"###;

#[tokio::test]
async fn all_fields() {
    assert_solving_snapshots!(
        "all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            id
            name
          }
        }
        "#
    );
}

#[tokio::test]
async fn nested_id_field_with_all_fields() {
    assert_solving_snapshots!(
        "nested_id_with_all_fields",
        SCHEMA,
        r#"
        query {
          products {
            price
            category
            name
            nested { id }
          }
        }
        "#
    );
}

#[tokio::test]
async fn single_field() {
    assert_solving_snapshots!(
        "single_field",
        SCHEMA,
        r#"
        query {
          products {
            price
          }
        }
        "#
    );
}
