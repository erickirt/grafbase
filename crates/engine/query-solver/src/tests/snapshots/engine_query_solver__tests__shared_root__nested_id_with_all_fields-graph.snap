---
source: crates/engine/query-solver/src/tests/shared_root.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Query.products\" ]\n    2 [ label = \"Product.price\" ]\n    3 [ label = \"Product.category\" ]\n    4 [ label = \"Product.name\" ]\n    5 [ label = \"Product.nested\" ]\n    6 [ label = \"Nested.id\" ]\n    7 [ label = \"Root#category\", shape=parallelogram, color=dodgerblue ]\n    8 [ label = \"products#category\", shape=box, color=dodgerblue ]\n    9 [ label = \"Root#name\", shape=parallelogram, color=dodgerblue ]\n    10 [ label = \"products#name\", shape=box, color=dodgerblue ]\n    11 [ label = \"Root#price\", shape=parallelogram, color=dodgerblue ]\n    12 [ label = \"products#price\", shape=box, color=dodgerblue ]\n    13 [ label = \"price#price\", shape=box, color=dodgerblue ]\n    14 [ label = \"nested#price\", shape=box, color=dodgerblue ]\n    15 [ label = \"id#price\", shape=box, color=dodgerblue ]\n    16 [ label = \"name#name\", shape=box, color=dodgerblue ]\n    17 [ label = \"nested#name\", shape=box, color=dodgerblue ]\n    18 [ label = \"id#name\", shape=box, color=dodgerblue ]\n    19 [ label = \"category#category\", shape=box, color=dodgerblue ]\n    20 [ label = \"nested#category\", shape=box, color=dodgerblue ]\n    21 [ label = \"id#category\", shape=box, color=dodgerblue ]\n    0 -> 1 [ label = \"\" ]\n    1 -> 2 [ label = \"\" ]\n    1 -> 3 [ label = \"\" ]\n    1 -> 4 [ label = \"\" ]\n    1 -> 5 [ label = \"\" ]\n    5 -> 6 [ label = \"\" ]\n    0 -> 7 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    0 -> 7 [ label = \"\", style=dashed,arrowhead=none ]\n    7 -> 8 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    8 -> 1 [ label = \"\", color=violet,arrowhead=none ]\n    0 -> 9 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    0 -> 9 [ label = \"\", style=dashed,arrowhead=none ]\n    9 -> 10 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    10 -> 1 [ label = \"\", color=violet,arrowhead=none ]\n    0 -> 11 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    0 -> 11 [ label = \"\", style=dashed,arrowhead=none ]\n    11 -> 12 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    12 -> 1 [ label = \"\", color=violet,arrowhead=none ]\n    12 -> 13 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    13 -> 2 [ label = \"\", color=violet,arrowhead=none ]\n    12 -> 14 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    14 -> 5 [ label = \"\", color=violet,arrowhead=none ]\n    14 -> 15 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    15 -> 6 [ label = \"\", color=violet,arrowhead=none ]\n    10 -> 16 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    16 -> 4 [ label = \"\", color=violet,arrowhead=none ]\n    10 -> 17 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    17 -> 5 [ label = \"\", color=violet,arrowhead=none ]\n    17 -> 18 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    18 -> 6 [ label = \"\", color=violet,arrowhead=none ]\n    8 -> 19 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    19 -> 3 [ label = \"\", color=violet,arrowhead=none ]\n    8 -> 20 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    20 -> 5 [ label = \"\", color=violet,arrowhead=none ]\n    20 -> 21 [ label = \"\", color=royalblue,fontcolor=royalblue ]\n    21 -> 6 [ label = \"\", color=violet,arrowhead=none ]\n}\n"
---
digraph {
    0 [ root]
    1 [ Query.products]
    2 [ Product.price]
    3 [ Product.category]
    4 [ Product.name]
    5 [ Product.nested]
    6 [ Nested.id]
    7 [ Root#category]
    8 [ products#category]
    9 [ Root#name]
    10 [ products#name]
    11 [ Root#price]
    12 [ products#price]
    13 [ price#price]
    14 [ nested#price]
    15 [ id#price]
    16 [ name#name]
    17 [ nested#name]
    18 [ id#name]
    19 [ category#category]
    20 [ nested#category]
    21 [ id#category]
    0 -> 1 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
    1 -> 3 [ label = "Field" ]
    1 -> 4 [ label = "Field" ]
    1 -> 5 [ label = "Field" ]
    5 -> 6 [ label = "Field" ]
    0 -> 7 [ label = "CreateChildResolver" ]
    0 -> 7 [ label = "HasChildResolver" ]
    7 -> 8 [ label = "CanProvide" ]
    8 -> 1 [ label = "Provides" ]
    0 -> 9 [ label = "CreateChildResolver" ]
    0 -> 9 [ label = "HasChildResolver" ]
    9 -> 10 [ label = "CanProvide" ]
    10 -> 1 [ label = "Provides" ]
    0 -> 11 [ label = "CreateChildResolver" ]
    0 -> 11 [ label = "HasChildResolver" ]
    11 -> 12 [ label = "CanProvide" ]
    12 -> 1 [ label = "Provides" ]
    12 -> 13 [ label = "CanProvide" ]
    13 -> 2 [ label = "Provides" ]
    12 -> 14 [ label = "CanProvide" ]
    14 -> 5 [ label = "Provides" ]
    14 -> 15 [ label = "CanProvide" ]
    15 -> 6 [ label = "Provides" ]
    10 -> 16 [ label = "CanProvide" ]
    16 -> 4 [ label = "Provides" ]
    10 -> 17 [ label = "CanProvide" ]
    17 -> 5 [ label = "Provides" ]
    17 -> 18 [ label = "CanProvide" ]
    18 -> 6 [ label = "Provides" ]
    8 -> 19 [ label = "CanProvide" ]
    19 -> 3 [ label = "Provides" ]
    8 -> 20 [ label = "CanProvide" ]
    20 -> 5 [ label = "Provides" ]
    20 -> 21 [ label = "CanProvide" ]
    21 -> 6 [ label = "Provides" ]
}
